plugins/reindex/src/test/java/org/elasticsearch/plugin/reindex/ReindexCornerCaseTests.java

```diff
@@ -0,0 +1,125 @@
+        ensureGreen();
+
+        indexRandom(true, client().prepareIndex("source", "test", "test").setSource("foo", "bar"));
+
+        // Copy the doc with the timestamp
+        ReindexRequestBuilder copy = reindex().source("source").destination("dest").refresh(true);
+        assertThat(copy.get(), responseMatcher().created(1));
+    }
+
+    public void testTimestamp() throws Exception {
+        copyDoc("{\"_timestamp\": {\"enabled\": true}}");
+        assertSearchHits(client().prepareSearch("dest").setQuery(existsQuery(TimestampFieldMapper.NAME)).get(), "test");
+        assertEquals(fetchTimestamp("source"), fetchTimestamp("dest"));
+    }
+
+    public void testTTL() throws Exception {
+        copyDoc("{\"_ttl\": {\"enabled\": true, \"default\": \"20d\"}}");
+        assertNotNull(client().prepareGet("dest", "test", "test").get().getField(TTLFieldMapper.NAME).getValue());
+    }
+
+    public void testRoutingCopiedByDefault() throws Exception {
+        routingTestCase(null, "bar");
+    }
+
+    public void testRoutingCopiedIfRequested() throws Exception {
+        routingTestCase("keep", "bar");
+    }
+
+    public void testRoutingDiscardedIfRequested() throws Exception {
+        routingTestCase("discard", null);
+    }
+
+    public void testRoutingSetIfRequested() throws Exception {
+        routingTestCase("=cat", "cat");
+    }
+
+    public void testRoutingSetIfWithDegenerateValue() throws Exception {
+        routingTestCase("==]", "=]");
+    }
+
+    /**
+     * Check that index-by-search does the right thing with routing.
+     *
+     * @param specification
+     *            if non-null then routing is specified as this on the
+     *            index-by-search request
+     * @param expectedRoutingAfterCopy
+     *            should the index-by-search request result in the routing being
+     *            copied (true) or stripped (false)
+     */
+    private void routingTestCase(String specification, String expectedRoutingAfterCopy) throws Exception {
+        indexRandom(true,
+                client().prepareIndex("source", "test", "has_routing").setRouting("bar").setSource("foo", "bar"));
+
+        assertNotNull(client().prepareGet("source", "test", "has_routing").setRouting("bar").get()
+                .getField(RoutingFieldMapper.NAME).getValue());
+
+        // Copy the child to a new type
+        ReindexRequestBuilder copy = reindex().source("source").destination("dest").refresh(true);
+        if (specification != null) {
+            copy.destination().setRouting(specification);
+        }
+        assertThat(copy.get(), responseMatcher().created(1));
+
+        // Make sure routing is intact on the copy
+        GetField routing = client().prepareGet("dest", "test", "has_routing").setRouting(expectedRoutingAfterCopy).get()
+                .getField(RoutingFieldMapper.NAME);
+        if (expectedRoutingAfterCopy == null) {
+            assertNull(expectedRoutingAfterCopy, routing);
```

Comment:
`expectedRoutingAfterCopy` is the assert message. Maybe add an explicit message or none at all instead of just "null"?